tools/syz-trace2syz/proggen/special_struct_handler.go

```diff
@@ -0,0 +1,39 @@
+// Copyright 2018 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package proggen
+
+import (
+	"github.com/google/syzkaller/pkg/log"
+	"github.com/google/syzkaller/prog"
+	"github.com/google/syzkaller/tools/syz-trace2syz/parser"
+)
+
+type structHandler func(syzType *prog.StructType, traceType parser.IrType, ctx *Context) parser.IrType
+
+var specialStructMap = map[string]structHandler{
+	"bpf_framed_program": bpfFramedProgramHandler,
+}
+
+func preprocessStruct(syzType *prog.StructType, traceType parser.IrType, ctx *Context) parser.IrType {
+	if structFunc, ok := specialStructMap[syzType.Name()]; ok {
+		return structFunc(syzType, traceType, ctx)
+	}
+	return traceType
+}
+
+func bpfFramedProgramHandler(syzType *prog.StructType, traceType parser.IrType, _ *Context) parser.IrType {
+	switch a := traceType.(type) {
+	case *parser.GroupType:
+		if a.Len > 1 {
+			straceStructArgs := make([]parser.IrType, len(syzType.Fields))
+			straceStructArgs[1] = a
+			straceArg0 := parser.GenDefaultIrType(syzType.Fields[0])
+			straceStructArgs[0] = straceArg0
+			straceStructArgs = append(straceStructArgs, parser.GenDefaultIrType(syzType.Fields[1]))
```

Comment:
The type description is:
```
bpf_framed_program {
	initr0	bpf_insn_init_r0
	body	array[bpf_insn]
	exit	bpf_insn_exit
} [packed]
```

This looks wrong in 2 aspects:
1. First we seem to need to use Fields[2] instead of Fields[1] here.
2. We already created array with 3 fields, filled [0] and [1], but instead of filling in [2], we instead append 4-th element.

I think we need something like:
```
straceStructArgs[0] = parser.GenDefaultIrType(syzType.Fields[0])
straceStructArgs[1] = a
straceStructArgs[2] = parser.GenDefaultIrType(syzType.Fields[2])
```

But on a more general note, it looks wrong to generate the framed program from unframed program. It may not pass validation afterwards. We already have a complete program, why do we need to frame it?

If we reorder the type to make raw the first field, can we then delete this whole file?
```
bpf_instructions [
	raw	array[bpf_insn]
	framed	bpf_framed_program
] [varlen]
```

Reply:
And delete GenDefaultIrType as well.